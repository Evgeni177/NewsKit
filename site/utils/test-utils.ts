import {
  renderToFragmentWithThemeFactory,
  renderWithThemeFactory,
} from 'newskit/test/test-utils';
import {docsThemeLight} from '../theme/doc-theme';

export const renderWithTheme = renderWithThemeFactory(docsThemeLight);

export const renderToFragmentWithTheme = renderToFragmentWithThemeFactory(
  docsThemeLight,
);
