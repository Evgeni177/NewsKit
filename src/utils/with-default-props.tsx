import React from 'react';
import {DeepPartial} from './types';
import {deepMerge} from './deep-merge';
import {getToken} from './get-token';
import {useTheme} from '../theme';

type PropsEvalFunction<P> = (props: P) => P;

export const withDefaultProps = <P extends {}>(
  Component: React.ComponentType<P>,
  defaultProps?: DeepPartial<P> | PropsEvalFunction<P>,
  defaultPresetsPath?: string,
): React.FC<P> => (props: P) => {
  const dProps =
    typeof defaultProps === 'function'
      ? (defaultProps as PropsEvalFunction<P>)(props)
      : defaultProps || {};
  const theme = useTheme();
  const overrides = getToken({theme}, defaultPresetsPath);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const {children, ...propsWithoutChildren} = props as any;
  return (
    <Component
      {...deepMerge(
        {} as P,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        {overrides} as any,
        dProps,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        propsWithoutChildren as any,
      )}
    >
      {children}
    </Component>
  );
};
