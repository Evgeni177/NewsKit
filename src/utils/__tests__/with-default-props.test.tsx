import React from 'react';
import {withDefaultProps} from '../with-default-props';
import {renderToFragmentWithTheme} from '../../test/test-utils';

describe('withDefaultProps', () => {
  const DummyComp: React.FC<any> = props => (
    <div>{JSON.stringify(props, null, 2)}</div>
  );

  test('will return a component with merged props taking a function as defaultProps', () => {
    const functionMock = jest.fn().mockReturnValue({functionMock: 'props'});
    const wrappedComponent = withDefaultProps(
      DummyComp,
      functionMock,
      'linkStandalone',
    );
    const props = {
      test: 'props',
    };

    const fragment = renderToFragmentWithTheme(wrappedComponent, props);

    expect(fragment).toMatchSnapshot();
    expect(functionMock).toHaveBeenCalledWith(props);
  });

  test('will return a component with merged props taking an object as defaultProps', () => {
    const fragment = renderToFragmentWithTheme(
      withDefaultProps(DummyComp, {test: 'prop'}, 'linkStandalone'),
    );

    expect(fragment).toMatchSnapshot();
  });

  test('will return a component with merged props taking undefined as defaultProps', () => {
    const fragment = renderToFragmentWithTheme(
      withDefaultProps(DummyComp, undefined, 'linkStandalone'),
    );

    expect(fragment).toMatchSnapshot();
  });
});
