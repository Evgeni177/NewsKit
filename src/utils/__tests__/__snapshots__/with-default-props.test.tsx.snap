// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withDefaultProps will return a component with merged props taking a function as defaultProps 1`] = `
<DocumentFragment>
  <div>
    {
  "overrides": {
    "stylePreset": "linkStandalone",
    "typographyPreset": "utilityLabel020"
  },
  "functionMock": "props",
  "test": "props"
}
  </div>
</DocumentFragment>
`;

exports[`withDefaultProps will return a component with merged props taking an object as defaultProps 1`] = `
<DocumentFragment>
  <div>
    {
  "overrides": {
    "stylePreset": "linkStandalone",
    "typographyPreset": "utilityLabel020"
  },
  "test": "prop"
}
  </div>
</DocumentFragment>
`;

exports[`withDefaultProps will return a component with merged props taking undefined as defaultProps 1`] = `
<DocumentFragment>
  <div>
    {
  "overrides": {
    "stylePreset": "linkStandalone",
    "typographyPreset": "utilityLabel020"
  }
}
  </div>
</DocumentFragment>
`;
